### PR TITLE
Fixed #13677 -- Propagated database choice from ModelFormSet queryset to ModelForm's choices.

### PR DESCRIPTION
**Pull request by rixx**
Moved the patch by @dgouldin to current Django.

Ticket link: https://code.djangoproject.com/ticket/13677

### CODE REVIEW

**Comment by rixx** [comment]
I'm not entirely in favor of this patch, I just moved it up from Django 1.2 - I think the use of `get_queryset()` might lead to performance issues, especially since it is called on every `BaseModelForm` creation.

**Comment by timgraham** [comment]
I don't think there are performance issues because `get_queryset()._db` doesn't evaluate the queryset. I would expect the `using` argument also be applied to `model.save()` to avoid the workaround described at https://stackoverflow.com/questions/37740848/use-a-specific-database-with-a-modelform? Documentation for the new argument is needed.

**Comment by timgraham** [comment]
To answer your questions on IRC:
1. to use the 'using' in 'save' calls, we need to keep it on the modelform - would you name it 'using' or '_using'?
I don't see a need for the underscore prefix.

2.  I'm not quite sure where I'm to document the new parameter – it'll definitely have to go into topics/forms/modelform.txt, but I haven't found a place in ref/ where it'd go. Is there something I overlooked?
The link for ModelForm is here: `docs/topics/forms/modelforms.txt:.. class:: ModelForm`. Preceding this change, that could be moved to `ref/forms/models.txt` and the `__init__()` parameters documented in some detail.

**Comment by rixx** [comment]
I added the `using` attribute, and I will add documentation next week. After that, I'll be offline for two weeks, and will respond to further comments after that.

**Comment by rixx** [comment]
Hah, "two weeks".

As there is no real place to add the `using` attribute, I added a short reference documentation of ``ModelForm``. As I am not a native speaker, I'd appreciate any comment on my phrasing there. Other than that I'd consider this PR ready for review. 👓

**Comment by rixx** [comment]
I rebased against current master. The test is no longer ready for merging due to a deprecation notice:

```
django.utils.deprecation.RemovedInDjango31Warning: `TestCase.multi_db` is deprecated. Databases available during this test can be defined using multiple_database.tests.ModelFormsetUsingTestCase.databases.
```

Could you point me at an example of a test case that was migrated from `multi_db` to `databases`? I'd be willing to apply that change here, too.

**Comment by timgraham** [comment]
Use `git log -S multi_db` to find the relevant commit (8c775391b78b2a4a2b57c5e89ed4888f36aada4b).

File: docs/ref/forms/models.txt [hunk 1]
```diff
@@ -81,3 +81,40 @@ Model Form API reference. For introductory material about model forms, see the
     the ``parent_model``, you must specify a ``fk_name``.
 
     See :ref:`inline-formsets` for example usage.
+
+``ModelForm``
```
Comment (timgraham) [inline docs/ref/forms/models.txt]: Unrelated documentation to the fix should go in a separate commit.

Comment (rixx) [reply docs/ref/forms/models.txt]: Is it unrelated? I was asked to add documentation for this feature, and since there was no API documentation for ModelForm yet, I included it.

**Comment by rixx** [comment]
Fixed the tests, separated the documentation into a separate commit. Should be good to go now.

File: tests/multiple_database/tests.py [hunk 2]
```diff
@@ -2107,3 +2107,29 @@ def test_reverse_one_to_one_relation(self):
...
+        class BookForm(forms.ModelForm):
+            class Meta:
+                model = Book
+                fields = ('title', 'authors', 'editor')
+        formset_class = modelformset_factory(Book, BookForm)
+        formset = formset_class(queryset=Book.objects.using('other').all())
+        for form in formset.forms:
+            databases = [i._state.db for i in form.fields['editor'].queryset]
+            databases.extend([i._state.db for i in form.fields['authors'].queryset])
+            self.assertEqual(set(databases), set(['other']))
```
Comment (felixxm) [inline tests/multiple_database/tests.py]: I think we should test the list of objects in `queryset` instead of their databases.

File: tests/multiple_database/tests.py [hunk 3]
```diff
@@ -2107,3 +2107,29 @@ def test_reverse_one_to_one_relation(self):
...
         with self.assertRaisesMessage(ValueError, self.router_prevents_msg):
             user.userprofile = profile
+
+
+class ModelFormsetUsingTestCase(TestCase):
+    databases = {'default', 'other'}
+
+    def test_formset_using(self):
+        from django import forms
+        from django.forms.models import modelformset_factory
```
Comment (felixxm) [inline tests/multiple_database/tests.py]: Is there any reason to import forms here instead of the top?

File: django/forms/models.py [hunk 4]
```diff
@@ -456,7 +464,7 @@ def save(self, commit=True):
             )
         if commit:
             # If committing, save the instance and the m2m data immediately.
-            self.instance.save()
+            self.instance.save(using=self.using)
```
Comment (felixxm) [inline django/forms/models.py]: Do we need this change? if yes then test is required.

File: tests/model_forms/models.py [hunk 5]
```diff
@@ -421,8 +421,8 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._savecount = 0
 
-    def save(self, force_insert=False, force_update=False):
-        super().save(force_insert, force_update)
+    def save(self, force_insert=False, force_update=False, using=None):
+        super().save(force_insert=force_insert, force_update=force_update, using=using)
```
Comment (felixxm) [inline tests/model_forms/models.py]: This change seems unrelated.

File: django/forms/models.py [hunk 6]
```diff
@@ -477,6 +485,8 @@ def modelform_factory(model, form=ModelForm, fields=None, exclude=None,
                       field_classes=None):
     """
     Return a ModelForm containing form fields for the given model.
+    You can optionally pass a ``form`` argument to use as a starting point for
+    constructing the ``ModelForm``.
```
Comment (felixxm) [inline django/forms/models.py]: I merged this comment in a387ef11171dc24ee6dbbcff7e76997677ad10dc.

**Comment by LuRsT** [comment]
Hey @rixx are you still working on this?

**Comment by felixxm** [comment]
Closing due to inactivity.
